core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -249,7 +249,14 @@ private[spark] class ExecutorAllocationManager(
    * yarn-client mode when AM re-registers after a failure.
    */
   def reset(): Unit = synchronized {
-    initializing = true
+    /**
+     * When some tasks need to be scheduled and initial executor = 0, resetting the initializing
+     * field may cause it to not be set to false in yarn.
+     * SPARK-20079: https://issues.apache.org/jira/browse/SPARK-20079
+     */
+    if (maxNumExecutorsNeeded() == 0) {
+      initializing = true
```

Comment:
In the original design of dynamic executor allocation,  this flag is set to "true" to avoid sudden executor number ramp up (because of first job submission) during initializing. You could check the comment.

And for the AM restart scenario, because all the executors will be re-spawned, this is similar to AM first start scenario (if the job is submitted during restart), so we set this flag to true.